compiler/rustc_codegen_llvm/src/context.rs

```diff
@@ -79,6 +79,10 @@ pub struct CodegenCx<'ll, 'tcx> {
     pub pointee_infos: RefCell<FxHashMap<(Ty<'tcx>, Size), Option<PointeeInfo>>>,
     pub isize_ty: &'ll Type,
 
+    /// Cache for the mapping from source index to llvm index for struct fields,
+    /// only present if synthetic fields are inserted for padding.
+    pub field_projection_cache: RefCell<FxHashMap<TyAndLayout<'tcx>, Vec<u32>>>,
```

Comment:
Hmm, storing it in the same cache definitely makes sense, but for now maybe just adding an `Option<Box<Vec<u32>>>` for the field projection would be enough?

As a tuple would be less self-documenting maybe replacing `lltypes` with `type_datas` like this:
```rs
pub struct CodegenCx<'ll, 'tcx> {
    // ...
    pub type_datas: RefCell<FxHashMap<(Ty<'tcx>, Option<VariantIdx>), TypeData<'ll>>>
    // ...
}

pub struct TypeData<'ll> {
    pub lltype: &'ll Type,
    pub field_projection: Option<Box<Vec<u32>>>,
}
```

Reply:
You can use `Option<Box<[u32]>>`, tho I suppose if it's too performance critical, `Box<Vec<` could also make sense.

Name-wise (tho I'm not too happy with any of my suggestions), I'd use `LoweredType` / `TypeLowering` instead of `TypeData`, and something like `field_remapping` / `field_renumbering` instead of `field_projection`.

Reply:
I went with the boxed slice, `TypeLowering` and `field_remapping`. The code is not performance critical. The performance regression in the `deep-vector-opt` test comes from thin LTO inexplicably taking a bit longer after the change.